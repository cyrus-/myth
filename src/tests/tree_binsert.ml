let s = "
type cmp =
  | CEq
  | CGt
  | CLt

type nat =
  | O
  | S of nat

type tree =
  | Leaf
  | Node of tree * nat * tree

let rec comp_nat (n1:nat) (n2:nat) : cmp =
  match n1 with
  | O -> (match n2 with
          | O -> CEq
          | S (n2) -> CLt)
  | S (n1) -> (match n2 with
              | O -> CGt
              | S (n2) -> comp_nat n1 n2)
;;

let tree_binsert : tree -> nat -> tree |>
  { Leaf => ( 0 => Node (Leaf, 0, Leaf)
            | 1 => Node (Leaf, 1, Leaf)
            | 2 => Node (Leaf, 2, Leaf) )
  | Node (Leaf, 1, Leaf) => ( 0 => Node (Node (Leaf, 0, Leaf), 1, Leaf)
                            | 1 => Node (Leaf, 1, Leaf)
                            | 2 => Node (Leaf, 1, Node (Leaf, 2, Leaf)) )
  | Node (Leaf, 0, Leaf) => ( 0 => Node (Leaf, 0, Leaf)
                            | 1 => Node (Leaf, 0, Node (Leaf, 1, Leaf))
                            | 2 => Node (Leaf, 0, Node (Leaf, 2, Leaf)) )
  | Node (Leaf, 2, Leaf) => ( 0 => Node (Node (Leaf, 0, Leaf), 2, Leaf)
                            | 1 => Node (Node (Leaf, 1, Leaf), 2, Leaf)
                            | 2 => Node (Leaf, 2, Leaf) )
  | Node (Node (Leaf, 0, Leaf), 1, Leaf) => ( 0 => Node (Node (Leaf, 0, Leaf), 1, Leaf)
					    | 1 => Node (Node (Leaf, 0, Leaf), 1, Leaf)
					    | 2 => Node (Node (Leaf, 0, Leaf), 1, Node(Leaf, 2, Leaf))
					    )
  | Node (Leaf, 0, Node (Leaf, 1, Leaf)) => ( 2 => Node (Leaf, 0, Node (Leaf, 1, Node(Leaf, 2, Leaf))) )
  | Node (Node (Leaf, 1, Leaf), 2, Leaf) => ( 0 => Node (Node (Node(Leaf, 0, Leaf), 1, Leaf), 2, Leaf) )
  | Node (Leaf, 1, Node (Leaf, 2, Leaf)) => ( 0 => Node (Node (Leaf, 0, Leaf), 1, Node (Leaf, 2, Leaf))
                                            | 1 => Node (Leaf, 1, Node (Leaf, 2, Leaf)) )
  | Node (Node (Leaf, 1, Leaf), 2, Leaf) => ( 0 => Node (Node (Node(Leaf, 0, Leaf), 1, Leaf), 2, Leaf) )
  } = ?"
