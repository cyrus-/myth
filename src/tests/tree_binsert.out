(* Generated Thu Nov 13 00:35:53 2014 *)
let tree_binsert : tree -> nat -> tree =
  let rec f1 (t1:tree) : nat -> tree =
    fun (n1:nat) ->
      match t1 with
        | Leaf -> Node (Leaf, n1, Leaf)
        | Node (t2, n2, t3) -> (match comp_nat n2 n1 with
                                  | CEq -> t1
                                  | CGt -> Node (f1 t2 n1, n2, t3)
                                  | CLt -> Node (t2, n2, f1 t3 n1))
  in
    f1
;;
